tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -2352,216 +2474,259 @@ private[optimizer] abstract class OptimizerCore(
     }
   }
 
-  private def foldBinaryOp(op: BinaryOp.Code, lhs: Tree, rhs: Tree)(
-      implicit pos: Position): Tree = {
+  private def foldBinaryOp(op: BinaryOp.Code, lhs: PreTransform,
+      rhs: PreTransform)(
+      implicit pos: Position): PreTransform = {
+    import BinaryOp._
+
+    (lhs, rhs) match {
+      case (PreTransLit(lhsLit), PreTransLit(rhsLit)) if op != String_+ =>
+        PreTransLit(constantFoldBinaryOp_except_String_+(op, lhsLit, rhsLit))
+
+      case _ =>
+        foldBinaryOpNonConstant(op, lhs, rhs)
+    }
+  }
+
+  private def foldBinaryOpNonConstant(op: BinaryOp.Code, lhs: PreTransform,
+      rhs: PreTransform)(
+      implicit pos: Position): PreTransform = {
     import BinaryOp._
-    @inline def default = BinaryOp(op, lhs, rhs)
+
+    @inline def default =
+      PreTransBinaryOp(op, lhs, rhs)
+
     (op: @switch) match {
       case === | !== =>
         val positive = (op == ===)
         (lhs, rhs) match {
-          case (lhs: Literal, rhs: Literal) =>
-            BooleanLiteral(literal_===(lhs, rhs) == positive)
+          case (_, PreTransTree(Null(), _)) if !lhs.tpe.isNullable =>
```

Comment:
Yes, it is, but ... how is that a problem here?

(I guess I could have used `PreTransLit(Null())` instead of `PreTransTree(Nul(), _)`, indeed. Probably a refactoring opportunity missed.)